### ISSUE TITLE
Add containerizingMode=NONE

### ISSUE THREAD

**Issue by rmannibucau** [GoogleContainerTools/jib#2788]
Hi,

Idea is to just skip the step to grab binaries/resources from current module.
This is particularly useful when creating a module taking a distro from another module or the same module which means just adding the the layers a folder.
Currently it works but there are some garbage in the final image.
Would be great to enable to clean that.

Romain

**Comment by chanseokoh** [comment]
Have you tried the filtering extension ([Maven](https://github.com/GoogleContainerTools/jib-extensions/tree/master/first-party/jib-layer-filter-extension-maven) / [Gradle](https://github.com/GoogleContainerTools/jib-extensions/tree/master/first-party/jib-layer-filter-extension-gradle))?

Otherwise, can you elaborate on your setup? Is it like an almost empty module with no Java sources and dependencies? What are you trying to achieve?

**Comment by rmannibucau** [comment]
@chanseokoh yes, tried but it workarounds the issue instead of solving it cleanly.
Setup is very common:

1. a phase or other module builds a folder with the app (can be a tomcat, tomee, karaf, batchee, ....)
2. jib module add that folder to the layer, period

so module is empty or, more generally, has config file only adding with additionalFiles and not exploded/packaged mode.

Indeed it is trivial to code with jib-core but since the effort is almost 0 I think it is worth to make plugin supporting it.
To be fully complete, and avoid 1 to require 2 modules in some cases (thinking to karaf), it  can be also needed to be able to filter project.getArtifacts() with an exclusion list and/or at least a scope.

I know it was not the original jib goal but I keep rewriting jib-like-maven-plugin just for that so hope it gets integrated.
Can do a PR if it helps.

**Comment by chanseokoh** [comment]
> 1. a phase or other module builds a folder with the app (can be a tomcat, tomee, karaf, batchee, ....)
> 2. jib module add that folder to the layer, period
> 
> so module is empty or, more generally, has config file only adding with additionalFiles and not exploded/packaged mode.

> Currently it works but there are some garbage in the final image.

Are you saying, e.g., a separate Jib-only sub-module (almost empty, sometimes with small config files) in a multi-module project put some files prepared by another sub-module only using the `<extraDirectories>` feature? In that case, what exactly are the "garbage files" that Jib puts but you want removed?

> I know it was not the original jib goal but I keep rewriting jib-like-maven-plugin just for that so hope it gets integrated.

Yeah, this feature sounds like a pretty low priority to us (even more so after we have the Jib plugin extension mechanism), I wouldn't have high hopes.

**Comment by rmannibucau** [comment]
@chanseokoh i want to remove the fake classpath/ folder jib enforces to create such image, without using filtering module (cause it is like saying "add that, remove that" which is a kind of anti-pattern - even if today it is the workaround I'm using).

**Comment by wouterhund** [comment]
I would like to see this as well. My use case is packaging a maven-built EAR into an image. Ideally I'd just be able to turn off the plugin's copying of dependencies and classes, and instead only using extraDirectories.

Using the filtering plugin is not an option for me because the Jib plugin aborts with:

> Obtaining project build output files failed; make sure you have compiled your project before trying to build the image.

Which happens because there is no built `.jar`, there is an `.ear`. I'm not expecting jib to support ear packaging, but it would be nice if I could use it to simply build a container by simply disabling some of its functionality.

**Comment by chanseokoh** [comment]
@wouterhund don't get your hopes too high, but I guess you have `<packaging>ear` in your `pom.xml`? I wonder if you can somehow suppress the error by creating a dummy `.java` file. The situation seems similar to [this](https://stackoverflow.com/a/66335872/1701388). Also maybe [Jib CLI](https://github.com/GoogleContainerTools/jib/tree/master/jib-cli) can be a fit?

**Comment by rmannibucau** [comment]
@wouterhund can you use this workaround https://rmannibucau.metawerx.net/jib-docker-layer-file-filtering.html (forced jar plugin def) ?

**Comment by wouterhund** [comment]
@chanseokoh Indeed I am using `<packaging>ear</packaging>` and am using the `maven-ear-plugin`.

This means that maven generates a `target/NAME.ear` file instead of a `.jar` file. So when jib is ran in PACKAGED mode it won't find the file as it's looking for a `.jar`. Oddly enough it seems that for ear packaged modules the compilation of java files is skipped entirely, even when invoking the `maven-compiler-plugin` explicitly, so there is no `target/classes` directory either which causes EXPLODED mode to fail as well.

I wanted to use the Jib CLI, which would be a great fit, but unfortunately ran into https://github.com/GoogleContainerTools/jib/issues/3183

I am able to explicitly invoke the maven-compiler-plugin using:

```
      <plugin>
        <artifactId>maven-compiler-plugin</artifactId>
        <executions>
          <execution>
            <id>default-compile</id>
            <phase>compile</phase>
            <goals>
              <goal>compile</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
```

Which satisfied jib, even though oddly enough the resulting `target/classes` directory is empty (so the compiler didn't pick up the dummy Java file). So this would give me a base I could work with using the filtering plugin.

------------------------

@rmannibucau Thanks for the tip! Unfortunately this doesn't seem to work when packaging is set to ear. Normally the maven-jar-plugin is skipped entirely, and when attaching its goal explicitly using:

```
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-jar-plugin</artifactId>
        <configuration>
          <skipIfEmpty>false</skipIfEmpty>
        </configuration>
        <executions>
          <execution>
            <id>jar</id>
            <phase>package</phase>
            <goals>
              <goal>jar</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
```

It errors with: `You have to use a classifier to attach supplemental artifacts to the project instead of replacing them.`

I might be able to work around that as well, but at that point it's getting a bit too hacky for my tastes.

------------------------

My current workaround is a tiny custom maven plugin which invokes jib-core to build the image exactly as I want to, but this is awkward because it doesn't integrate with the rest of jib's plugin configuration such as credential management. Unfortunately the `jib-plugins-common` module is not available on Nexus, so a few things has to be copy pasted over which I'm not happy with either.

So I will investigate if switching back to the `jib-maven-plugin` using an empty `target/classes` and filtering to remove undesired files works out, but it would be very nice if the plugin could offer a containerizingMode which essentially does nothing to support these kind of use cases.

**Comment by chanseokoh** [comment]
Thanks for all the great details. This helps a lot.

> So when jib is ran in PACKAGED mode it won't find the file as it's looking for a `.jar`.

Then I guess another hack for PACKAGED would be to create a dummy `.jar` file and remove it later using the filter extension. It's still very hacky though. 🤣

As you said, Jib CLI seems like a good fit in this case. We definitely want to improve the usability of Jib CLI, and we'll work on #3183. @mpeddada1

**Comment by rmannibucau** [comment]
@wouterhund you can always attach the ear with a classifier or just use another module to make the image - but still a workaround

@chanseokoh jib cli must not be seen as an alternative for maven build or so and I would expect jib-cli to be able to do the same as gradle task and maven plugin but not more - or reversed, I would expect the build plugins to be able to do as much as jib cli. I have a few cases which are not "simple" flat java classpath - embedding distro generally - and I use a ton of workarounds (from the filter ext to the complete exec-maven-plugin with jib-core). Would be neat to make jib exposing its actual underlying features and if well done, CLI, maven plugin and gradle task would be generated from jib-core API so no maintenance cost ;).

**Comment by mpeddada1** [comment]
> As you said, Jib CLI seems like a good fit in this case. We definitely want to improve the usability of Jib CLI, and we'll work on #3183. @mpeddada1

Update: Jib CLI 0.6.0 has been released with this feature (#3183).
